File: bindings/python/src/anki_forge_python/structured.py
from __future__ import annotations

import json

from .contracts import ContractValidationError, validate_contract_payload
from .errors import ProtocolParseError, RuntimeInvocationError
from .helpers import helper_view
from .raw import run_raw


def run_structured(command: str, request: dict, **runtime_kwargs) -> dict:
    raw = run_raw(command, request, **runtime_kwargs)

    if raw.exit_status != 0:
        raise RuntimeInvocationError(
            f"{command} exited with status {raw.exit_status}",
            command=command,
            exit_status=raw.exit_status,
            stdout=raw.stdout,
            stderr=raw.stderr,
            resolved_runtime=raw.resolved_runtime,
            failure_phase="process-exit",
        )

    try:
        parsed = json.loads(raw.stdout)
    except json.JSONDecodeError as error:
        raise ProtocolParseError(
            str(error),
            command=command,
            exit_status=raw.exit_status,
            stdout=raw.stdout,
            stderr=raw.stderr,
            resolved_runtime=raw.resolved_runtime,
            parse_phase="json",
        ) from error

    try:
        validate_contract_payload(command, parsed)
    except ContractValidationError as error:
        raise ProtocolParseError(
            str(error),
            command=command,
            exit_status=raw.exit_status,
            stdout=raw.stdout,
            stderr=raw.stderr,
            resolved_runtime=raw.resolved_runtime,
            parse_phase=error.parse_phase,
        ) from error

    result = dict(parsed)
    result["resolved_runtime"] = raw.resolved_runtime
    result["raw_command"] = {
        "command": raw.command,
        "argv": raw.argv,
        "exit_status": raw.exit_status,
    }
    result["helper"] = helper_view(command, parsed, request)
    return result


def normalize(request: dict, **runtime_kwargs) -> dict:
    return run_structured("normalize", request, **runtime_kwargs)


def build(request: dict, **runtime_kwargs) -> dict:
    return run_structured("build", request, **runtime_kwargs)


def inspect(request: dict, **runtime_kwargs) -> dict:
    return run_structured("inspect", request, **runtime_kwargs)


def diff(request: dict, **runtime_kwargs) -> dict:
    return run_structured("diff", request, **runtime_kwargs)

File: bindings/node/src/structured.js
import fs from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';

import { ProtocolParseError, RuntimeInvocationError } from './errors.js';
import { validateContractPayload } from './contracts.js';
import { helperView } from './helpers.js';
import { runRaw } from './raw.js';

async function runStructured(command, request, runtimeOptions) {
  const raw = await runRaw(command, request, runtimeOptions);

  if (raw.exitStatus !== 0 && command !== 'product-build') {
    throw new RuntimeInvocationError(`${command} exited with status ${raw.exitStatus}`, {
      command,
      exitStatus: raw.exitStatus,
      stdout: raw.stdout,
      stderr: raw.stderr,
      resolvedRuntime: raw.resolvedRuntime,
      failurePhase: 'process-exit',
    });
  }

  let parsed;
  try {
    parsed = JSON.parse(raw.stdout);
  } catch (error) {
    if (raw.exitStatus !== 0) {
      throw new RuntimeInvocationError(`${command} exited with status ${raw.exitStatus}`, {
        command,
        exitStatus: raw.exitStatus,
        stdout: raw.stdout,
        stderr: raw.stderr,
        resolvedRuntime: raw.resolvedRuntime,
        failurePhase: 'process-exit',
      });
    }
    throw new ProtocolParseError(error.message, {
      command,
      exitStatus: raw.exitStatus,
      stdout: raw.stdout,
      stderr: raw.stderr,
      resolvedRuntime: raw.resolvedRuntime,
      parsePhase: 'json',
    });
  }

  try {
    validateContractPayload(command, parsed);
  } catch (error) {
    if (raw.exitStatus !== 0) {
      throw new RuntimeInvocationError(`${command} exited with status ${raw.exitStatus}`, {
        command,
        exitStatus: raw.exitStatus,
        stdout: raw.stdout,
        stderr: raw.stderr,
        resolvedRuntime: raw.resolvedRuntime,
        failurePhase: 'process-exit',
      });
    }
    throw new ProtocolParseError(error.message, {
      command,
      exitStatus: raw.exitStatus,
      stdout: raw.stdout,
      stderr: raw.stderr,
      resolvedRuntime: raw.resolvedRuntime,
      parsePhase: error.parsePhase ?? 'contract-shape',
    });
  }

  return {
    ...parsed,
    resolvedRuntime: raw.resolvedRuntime,
    rawCommand: {
      command: raw.command,
      argv: raw.argv,
      exitStatus: raw.exitStatus,
    },
    helper: helperView(command, parsed, request),
  };
}

export function normalize(request, runtimeOptions = {}) {
  return runStructured('normalize', request, runtimeOptions);
}

export function build(request, runtimeOptions = {}) {
  return runStructured('build', request, runtimeOptions);
}

async function runProductBuild(request, runtimeOptions) {
  const tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'anki-forge-node-product-'));
  try {
    let inputPath = request.inputPath;
    if (request.productDocument !== undefined) {
      inputPath = path.join(tempDir, 'project.product-v3.json');
      await fs.writeFile(inputPath, JSON.stringify(request.productDocument), 'utf8');
    }
    if (!inputPath) {
      throw new TypeError('productBuild requires inputPath or productDocument');
    }
    return await runStructured(
      'product-build',
      {
        ...request,
        inputPath,
        apkgOut: request.apkgOut ?? path.join(tempDir, 'deck.apkg'),
      },
      runtimeOptions,
    );
  } finally {
    await fs.rm(tempDir, { recursive: true, force: true });
  }
}

export async function productBuild(request, runtimeOptions = {}) {
  if (!request.apkgOut) {
    throw new TypeError('productBuild requires apkgOut so the built artifact can be retained');
  }
  return await runProductBuild(request, runtimeOptions);
}

export async function productValidate(request, runtimeOptions = {}) {
  const result = await runProductBuild(request, runtimeOptions);
  return { ...result, artifact: null };
}

export function templateValidate(request, runtimeOptions = {}) {
  return productValidate(request, runtimeOptions);
}

export function inspect(request, runtimeOptions = {}) {
  return runStructured('inspect', request, runtimeOptions);
}

export function diff(request, runtimeOptions = {}) {
  return runStructured('diff', request, runtimeOptions);
}
